File: app/src/data/e2ee.ts
import "react-native-get-random-values"; // polyfills crypto.getRandomValues for tweetnacl
import nacl from "tweetnacl";
import util from "tweetnacl-util";

// Byte-for-byte compatible with daemon/e2ee.py (PyNaCl Box == tweetnacl box):
//   key exchange Curve25519, cipher XSalsa20-Poly1305,
//   frame = base64( nonce(24) || ciphertext(+16 tag) ).
// So a frame sealed here opens on the daemon and vice-versa.

export function generateKeyPair() {
  const kp = nacl.box.keyPair();
  return { pub: util.encodeBase64(kp.publicKey), sec: util.encodeBase64(kp.secretKey) };
}

export function seal(plaintext: string, mySecB64: string, peerPubB64: string): string {
  const nonce = nacl.randomBytes(nacl.box.nonceLength);
  const ct = nacl.box(util.decodeUTF8(plaintext), nonce, util.decodeBase64(peerPubB64), util.decodeBase64(mySecB64));
  const frame = new Uint8Array(nonce.length + ct.length);
  frame.set(nonce);
  frame.set(ct, nonce.length);
  return util.encodeBase64(frame);
}

export function open(frameB64: string, mySecB64: string, peerPubB64: string): string {
  const frame = util.decodeBase64(frameB64);
  const nonce = frame.slice(0, nacl.box.nonceLength);
  const ct = frame.slice(nacl.box.nonceLength);
  const msg = nacl.box.open(ct, nonce, util.decodeBase64(peerPubB64), util.decodeBase64(mySecB64));
  if (!msg) throw new Error("decrypt failed (key mismatch or tamper)");
  return util.encodeUTF8(msg);
}

File: spine/comms/e2ee.py
# -*- coding: utf-8 -*-
"""End-to-end encryption for the relay path - byte-for-byte compatible with
Paseo's scheme (packages/surfaces/relay/src/crypto.ts) so the wire format is a known
quantity:

  * key exchange : Curve25519  (NaCl crypto_box / box.before)
  * encryption   : XSalsa20-Poly1305  (NaCl crypto_box / box.after)
  * frame        : [nonce (24 bytes)][ciphertext(+16-byte tag)]

PyNaCl's Box == NaCl crypto_box == tweetnacl's box, so a frame sealed here opens
in the phone's tweetnacl/libsodium and vice-versa. The relay only ever sees the
frame bytes (opaque) - it cannot read or forge traffic."""
import base64
from nacl.public import PrivateKey, PublicKey, Box
from nacl.utils import random as _random

NONCE = Box.NONCE_SIZE  # 24


def generate_keypair():
    sk = PrivateKey.generate()
    return sk, sk.public_key


def export_pub(pk):
    return base64.b64encode(bytes(pk)).decode()


def import_pub(b64):
    return PublicKey(base64.b64decode(b64))


def export_sec(sk):
    return base64.b64encode(bytes(sk)).decode()


def import_sec(b64):
    return PrivateKey(base64.b64decode(b64))


def seal(plaintext, my_sk, peer_pk, nonce=None):
    """plaintext bytes -> frame bytes (nonce||ciphertext)."""
    box = Box(my_sk, peer_pk)
    if nonce is None:
        nonce = _random(NONCE)
    return bytes(box.encrypt(plaintext, nonce))   # PyNaCl prepends the nonce


def open_frame(frame, my_sk, peer_pk):
    """frame bytes (nonce||ciphertext) -> plaintext bytes; raises on tamper."""
    return Box(my_sk, peer_pk).decrypt(frame)


def seal_b64(plaintext, my_sk, peer_pk):
    return base64.b64encode(seal(plaintext, my_sk, peer_pk)).decode()


def open_b64(frame_b64, my_sk, peer_pk):
    return open_frame(base64.b64decode(frame_b64), my_sk, peer_pk)
